contracts/Curation.sol

```diff
@@ -0,0 +1,529 @@
+        bytes4 spec = bytes4(0x01);
+        bytes4 calldataLength = bytes4(0x24); // 4 + 32
+        bytes4 signature = this.resolveChallenge.selector;
+        bytes memory executionScript = new bytes(scriptLength);
+        // concatenate spec + address(this) + calldataLength + calldata
+        // TODO: should we put this somewhere in aragonOS to be able to reuse it? (if it's not already there)
+        assembly {
+            mstore(add(executionScript, 0x20), spec)
+            mstore(add(executionScript, 0x24), mul(address, exp(2,96)))
+            mstore(add(executionScript, 0x38), calldataLength)
+            mstore(add(executionScript, 0x3C), signature)
+            mstore(add(executionScript, 0x40), entryId)
+        }
+
+        uint256 voteId = voting.newVote(executionScript, "");
+
+        challenges[entryId] = Challenge({
+            challenger: msg.sender,
+            date: getTimestamp(),
+            amount: amount,
+            lockId: lockId,
+            voteId: voteId,
+            dispensationPct: dispensationPct
+        });
+
+        votes[voteId] = Vote({
+            closed: false,
+            result: false,
+            totalWinningStake: 0,
+            votersRewardPool: 0
+        });
+
+        NewChallenge(entryId, msg.sender);
+
+        return entryId;
+    }
+
+    /**
+     * @notice Resolve Challenge for entry with id `entryId`
+     * @param entryId Id of the Application/Challenge
+     */
+    function resolveChallenge(bytes32 entryId) isInitialized public {
+        require(challengeExists(entryId));
+        Challenge storage challenge = challenges[entryId];
+        Application storage application = applications[entryId];
+        Vote storage vote = votes[challenge.voteId];
+
+        require(voting.isClosed(challenge.voteId));
+        vote.closed = true;
+        (vote.result, vote.totalWinningStake,) = voting.getVoteResult(challenge.voteId);
+
+        address winner;
+        address loser;
+        uint256 loserLockId;
+        uint256 amount;
+        if (vote.result == false) { // challenge not accepted (application remains)
+            winner = application.applicant;
+            loser = challenge.challenger;
+            loserLockId = challenge.lockId;
+            amount = challenge.amount;
+
+            // it's still in application phase (not registered yet)
+            if (!application.registered) {
+                application.registered = true;
+                // insert in Registry app
+                registry.add(application.data);
+            }
+        } else { // challenge accepted (application rejected)
+            winner = challenge.challenger;
+            loser = application.applicant;
+            loserLockId = application.lockId;
+            amount = application.amount;
+
+            // Remove from Registry app
+            application.registered = false;
+            registry.remove(entryId);
```

Comment:
We should have special logic for whether the application was already in the registry or the challenge is done during the `applyStageLen` period. This isn't causing a crash because the registry doesn't care whether an entry was present in the registry before attempting to delete it: https://github.com/aragonlabs/registry/pull/3/files#r207185327

Reply:
Yes, actually I think that's the reason I was not doing those checks on Registry app: it was making things easier here, as the operation would become "idempotent", i.e., it doesn't matter what's the state, the result will always be the desired (in this case, having the entry out of the registry).